dattri/algorithm/influence_function.py

```diff
@@ -815,3 +815,187 @@ def _get_layer_wise_reps(self,
+                used to calculate the train/test representations. If None, all linear
+                modules are used. This should be a string or a list of strings if
+                multiple modules are needed. The name of module should follow the
+                key of model.named_modules(). Default: None.
+            device (str): Device to run the attributor on. Default is "cpu".
+            damping (float): Damping factor used for non-convexity in EK-FAC IFVP
+                calculation. Default is 0.0.
+
+        Raises:
+            ValueError: If there are multiple checkpoints in `task`.
+        """
+        super().__init__(task, None, device)
+        if len(self.task.checkpoints) > 1:
+            error_msg = ("Received more than one checkpoint. "
+                         "Ensemble of EK-FAC is not supported.")
+            raise ValueError(error_msg)
+
+        if not module_name:
+            # Select all linear layers by default
+            module_name = [
+                name for name, mod in self.task.model.named_modules()
+                if isinstance(mod, torch.nn.Linear)
+            ]
+        if not isinstance(module_name, list):
+            module_name = [module_name]
+
+        self.module_name = module_name
+
+        # Update layer_name corresponding to selected modules
+        self.layer_name = [name + ".weight" for name in self.module_name]
+
+        self.damping = damping
+        self.name_to_module = {
+            name: self.task.model.get_submodule(name) for name in module_name
+        }
+        self.module_to_name = {v: k for k, v in self.name_to_module.items()}
+
+        self.layer_cache = {}  # cache for each layer
+
+        def _ekfac_hook(module: torch.nn.Module,
+                        inputs: Union[Tensor, Tuple[Tensor]],
+                        outputs: Union[Tensor, Tuple[Tensor]],
+            ) -> None:
+            # Unpack tuple outputs if necessary
+            if isinstance(inputs, tuple):
+                inputs = inputs[0]
+
+            if isinstance(outputs, tuple):
+                outputs = outputs[0]
+
+            outputs.retain_grad()
+            name = self.module_to_name[module]
+            # Cache the inputs and outputs
+            self.layer_cache[name] = (inputs, outputs)
+
+        self.handles = []
+        for name in module_name:
+            mod = task.model.get_submodule(name)
+            self.handles.append(mod.register_forward_hook(_ekfac_hook))
+
+    def cache(
+        self,
+        full_train_dataloader: DataLoader,
+        max_iter: Optional[int] = None,
+    ) -> None:
+        """Cache the dataset and statistics for inverse hessian/fisher calculation.
+
+        Cache the full training dataset as other attributors.
+        Estimate and cache the covariance matrices, eigenvector matrices
+        and corrected eigenvalues based on the distribution of training data.
```

Comment:
based on the distribution of training data -> based on samples of training data.